assets/js/base/hooks/use-url-query-string.js

```diff
@@ -0,0 +1,137 @@
+/**
+ * External dependencies
+ */
+import { useEffect, useReducer } from '@wordpress/element';
+import { addQueryArgs, getQueryArg } from '@wordpress/url';
+import { useQueryStateContext } from '@woocommerce/base-context/query-state-context';
+import isShallowEqual from '@wordpress/is-shallow-equal';
+
+const hasWindowDependencies =
+	typeof window === 'object' &&
+	window.hasOwnProperty( 'history' ) &&
+	window.hasOwnProperty( 'location' ) &&
+	typeof window.addEventListener === 'function' &&
+	typeof window.removeEventListener === 'function';
+
+const update = ( urlKey, urlValue ) => {
+	return {
+		type: 'update',
+		urlKey,
+		urlValue,
+	};
+};
+
+const reset = ( initializationValues ) => {
+	return {
+		type: 'reset',
+		initializationValues,
+	};
+};
+
+const reducer = ( state, action ) => {
+	const { type, urlKey, urlValue, initializationValues } = action;
+	switch ( type ) {
+		case 'reset':
+			state = initializationValues || state;
+			break;
+		case 'update':
+			state =
+				! state[ urlKey ] ||
+				! isShallowEqual( state[ urlKey ], urlValue )
+					? { ...state, [ urlKey ]: urlValue }
+					: state;
+			break;
+		default:
+			throw new Error( 'Invalid action type for reducer.' );
+	}
+	return state;
+};
+
+const initializeState = ( { urlKeys, queryStateContext } ) => {
```

Comment:
Minor nitpick, let's rename `queryStateContext` to `suffix` or something so this function is generic and reinforce that the func doesn't know about our context.